packages/devtools_app/lib/src/debugger/flutter/console.dart

```diff
@@ -58,133 +50,22 @@ class _DebuggerConsoleState extends State<DebuggerConsole> {
-/// top-right corner.
-// TODO(ditman): Reuse this in `logging/flutter/logging_screen.dart`?
-class Console extends StatelessWidget {
-  const Console({
-    this.controls = const <Widget>[],
-    this.lines = const <String>[],
-  }) : super();
-
-  final List<Widget> controls;
-  final List<String> lines;
-
-  @override
-  Widget build(BuildContext context) {
-    return Stack(
-      children: [
-        _ConsoleOutput(lines: lines),
-        if (controls.isNotEmpty)
-          _ConsoleControls(
-            controls: controls,
-          ),
-      ],
-    );
-  }
-}
-
-/// Renders a top-right aligned ButtonBar wrapping a List of IconButtons
-/// (`controls`).
-class _ConsoleControls extends StatelessWidget {
-  const _ConsoleControls({
-    this.controls,
-  }) : super();
-
-  final List<Widget> controls;
-
-  @override
-  Widget build(BuildContext context) {
-    return Container(
-      alignment: Alignment.topRight,
-      child: ButtonBar(
-        buttonPadding: EdgeInsets.zero,
-        alignment: MainAxisAlignment.end,
-        children: controls,
-      ),
-    );
-  }
-}
-
-/// Renders a widget with the output of the console.
-///
-/// This is a ListView of text lines, with a monospace font and a border.
-class _ConsoleOutput extends StatefulWidget {
-  const _ConsoleOutput({
-    Key key,
-    this.lines,
-  }) : super(key: key);
-
-  final List<String> lines;
-
-  @override
-  _ConsoleOutputState createState() => _ConsoleOutputState();
-}
-
-class _ConsoleOutputState extends State<_ConsoleOutput> {
-  // The scroll controller must survive ConsoleOutput re-renders
-  // to work as intended, so it must be part of the "state".
-  final ScrollController _scroll = ScrollController();
-
-  // Whenever the widget updates, refresh the scroll position if needed.
-  @override
-  void didUpdateWidget(oldWidget) {
-    if (_scroll.hasClients && _scroll.atScrollBottom) {
-      _scroll.autoScrollToBottom();
-    }
-    super.didUpdateWidget(oldWidget);
-  }
-
-  @override
-  Widget build(BuildContext context) {
-    final theme = Theme.of(context);
-    final textStyle =
-        theme.textTheme.bodyText2.copyWith(fontFamily: 'RobotoMono');
-
-    return OutlineDecoration(
-      child: Scrollbar(
-        child: ListView.builder(
-          padding: const EdgeInsets.all(denseSpacing),
-          itemCount: widget.lines.length,
-          itemExtent: CodeView.rowHeight,
-          controller: _scroll,
-          itemBuilder: (context, index) {
-            return RichText(
-              text: TextSpan(
-                children: processAnsiTerminalCodes(
-                  widget.lines[index],
-                  textStyle,
-                ),
-              ),
-              maxLines: 1,
-            );
-          },
+    final numLines = _lines.length;
+    final disabled = numLines == 0;
+    return Console(
+      lines: _lines,
+      controls: [
+        CopyToClipboardControl(
+          data: disabled ? null : _lines.join('\n'),
```

Comment:
This will join the lines each time this widget is rebuilt. I'm not sure if that's a performance consideration or not; the previous code wouldn't join the lines until the copy action was invoked.

Reply:
We could also have the `DebuggerController.stdio` API provide raw text (a String or StringBuilder) if we thought that would be easier for consumers.

Reply:
Your suggestion of making `data` a `dataProvider` function is going to prevent this issue. About returning a raw text or not, I'm not sure, unless we need to do smarter things in the process of converting this to a string?

Reply:
Making it lazy should be enough; in the future we may want to look at returning the raw char data to clients, as well as a pre-computed set of lines, so clients that want to operate on the unprocessed data can. Definitely not a today issue.

Reply:
Done!